vmware: Support external customer VMs

joker-at-work opened the pull request:
Please see the individual commits' messages.

joker-at-work committed:
vmware: Expose external customer traits

With splitting up clusters into internal and external workloads using
DRS HostGroups, we have to expose information to the scheduler which
hosts can support external workloads. This could be done manually by
adding the trait defined in
`nova.utils.EXTERNAL_CUSTOMER_SUPPORTED_TRAIT` to hosts, but that seems
cumbersome and error-prone. Therefore, we detect HostGroups
automatically through a naming convention and, if we find some, expose
appropriate traits for them.

We expose traits per HostGroup so we can in the future also support
customer-owned HVs where all VMs in a domain should only run on hosts
dedicated to that customer. Having traits exposed for them allows us to
filter in the scheduler.

We do not use the `DRS_PREFIX` for finding appropriate HostGroup
objects, because these are managed by an external entity and not by Nova
and thus should not be named with Nova's prefix.

We refresh the HostGroups every time we refresh the cluster resources.
We also save the HostGroups in the `VCState`'s stats, so we can use them
later on e.g. if we want to scheduler manually instead of via DRS and
when checking which VmGroup a VM should be put to during
spawning/migrating.

Change-Id: Ic7da95442c99cc51798d3555325d477a0c4ce5ea

seb-kro commented on nova/virt/vmwareapi/driver.py:
```diff
@@ -611,6 +611,40 @@ def update_provider_tree(self, provider_tree, nodename, allocations=None):
+                    continue
+
+                trait = utils.EXTERNAL_CUSTOMER_DOMAIN_TRAIT.format(
+                    hg_name.upper().replace('-', '_'))
```
Are we positive that `hg_prefix` will not include any `-`? Otherwise prefix matching would fail for the existing traits. I assume that `external_customer_domain_name_prefixes` only includes `iaas-` for now, which is fine due to the suffix removal. But for avoiding future issues could still make sense to validate and alert here?

joker-at-work replied:
I think it would be fine. We will support hostgroups named `iaas-whatever` and if we ever have a new domain named "more-iaas-whatever" we add the prefix "more-iaas-" and expect a HostGroup named "more-iaas". I think I didn't get the part where that's a problem with traits. Could you give an example?

seb-kro replied:
Let's say we have prefix `more-iaas-` and host group `more-iaas-xyz`. I believe this would add `CUSTOM_EXTERNAL_CUSTOMER_MORE_IAAS_XYZ` to `wanted_traits`. But looks for traits that start with `CUSTOM_EXTERNAL_CUSTOMER_MORE-IAAS` to add to `existing_traits`. Note that here the `-` in `MORE-IAAS` is kept. So even if it exists, `CUSTOM_EXTERNAL_CUSTOMER_MORE_IAAS_XYZ` would not be added to `existing_traits`.

seb-kro replied:
Although it probably makes a lot more sense to simply replace `-` with `_`  also for the matching of existing traits rather than to validate or alert about anything.

joker-at-work replied:
So you mean in line 623 we should add a `.replace('-', '_')`? yeah, sounds about right

joker-at-work replied:
nice find.

joker-at-work committed:
vmware: Add helper to update VmGroup members

We will in need to update VmGroup members of more than the admin group
in the future for external customers. Since updating in parallel can be
bad, we add a lock into the function. If the VmGroup does not exist, we
raise an exception so the surrounding code can handle the situation.

Change-Id: I51cfecc0a7fb27b4e59561ca93dddb4f6cc6bcd9

joker-at-work committed:
vmware: Extract create_vm_host_rule() helper

We didn't previously need this code outside of
`_create_cluster_group_rules_spec()` but with supporting external
customers, we will have to create more host rules.

Change-Id: I16aa0e04701e17e8225053983ef40619c18b9899

joker-at-work committed:
vmware: Extract Instance filtering from sync_server_group()

We need the same filtering in updating/syncing the VmGroups we use to
support external customers, so we make the functionality available
outside `sync_server_group()` and a little more generic.

Change-Id: Icb53377b8fa2e31aa4bfd31020366106b6f00bfc

joker-at-work committed:
vmware: Support external customer VmGroups

For keeping external customer VMs on specific hosts, we get some DRS
HostGroups created by the operations team and Nova creates DRS affinity
and anti-affinity rules to keep VMs on/away from these host.

We implement 2 methods here: One for creating the appropriate VmGroups
and DRS rules and one for syncing the current members for a VmGroup.
These methods get called during VM operations through
`update_cluster_placement()` which we already use to sync server-groups
and the admin VmGroup (for nova-bigvm-enabled regions) into the cluster.

I opted for syncing all members of a VmGroup instead of adding/removing
one VM, because we plan to run a sync-loop which can then re-use the
code.

Keeping the VmGroup and DRS rule cleanup hidden behind an extra argument
was meant to keep the number of reconfigures minimal during VM
operations. The planned sync-loop can do the clean-up.

We take a lock per VmGroup and re-fetch the Instances inside the lock to
reduce the possibility for race-conditions.

Change-Id: Ia065958822b636c35b84349f44e88f38c923acbb

joker-at-work committed:
vmware: Add sync-loops for external customer VMs

We want to make sure that external customer VMs run on the dedicated
hosts. In case a VM - for whatever reason - is not part of the VmGroup,
we will catch that in the sync-loop and add it back.

Additionally, we run the cleanup part of the DRS rules sync, which will
get rid of DRS rules and VmGroups for no-longer-existing or empty
HostGroups.

Change-Id: I2cdb9f6823b7254197662440c011f6e1f7abd969

seb-kro commented on nova/conf/vmware.py:
```diff
@@ -515,6 +515,34 @@
 syncing the traits. This setting defines how long to sleep between runs.
 
+Possible values:
+ * integer >= time in seconds to sleep between runs
```
Probably meant to be `* integer >= 0: ...`?

seb-kro commented on nova/conf/vmware.py:
```diff
@@ -515,6 +515,34 @@
+no-longer-necessary ones. This setting defines how long to sleep between runs.
+
+Possible values:
+ * integer >= time in seconds to sleep between runs
```
Probably meant to be `* integer >= 0: ...`?

seb-kro commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -1273,6 +1276,233 @@ def update_admin_vm_group_membership(self, instance, remove=False):
+            # This cluster does not manage external customer VMs.
+            return
+
+        self.sync_external_customer_drs_rules()
```
Is it intended to run this with the default argument of `do_cleanup=True`? Since the sync loop does cleanup anyways.

joker-at-work replied:
No! Good catch. I think I even said so in the commit message.

seb-kro commented on nova/virt/vmwareapi/driver.py:
```diff
@@ -981,6 +1017,61 @@ def _custom_traits_sync_loop(self, compute_host):
+        We pick the list of existing VmGroup objects from the cluster and sync
+        the members for them. _external_customer_drs_rules_sync_loop() takes
+        care of creating/deleting VmGroup objects for existing HostGroups.
+        """
```
Looks like `_external_customer_drs_rules_sync_loop` creates missing VmGroups, but does not delete any redundant groups? So this would still need to be added.

joker-at-work replied:
hm ... I never heard of redundant VmGroups being created before - if you mean duplicates, so I hadn't planned to do so. If, by chance, you meant no longer necessary, then I do see code for it through the `do_cleanup=True`.

seb-kro replied:
My thinking was more like this: Since a new host group is created for each external customer, I would also expect that host group to be deleted when a customer is no longer using our product. Also host groups might be created with a wrong name accidentally / for temporary testing / ... and then be removed.

In any of these scenarios, this code will so far create DRS rules for new host groups and delete the corresponding rules when a host group is removed. I.e., DRS rules are always kept in sync.

So I assumed we would do the same for the VmGroups, instead of keeping the deprecated VmGroup when a host group is removed for whichever reason.

joker-at-work replied:
Ah, I missed that part. Thank you.

seb-kro commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -1273,6 +1276,233 @@ def update_admin_vm_group_membership(self, instance, remove=False):
+            if not expected_members:
+                LOG.info("Sync for external customer VmGroup %s done: "
+                         "No expected members.", vm_group_name)
+                return
```
Given that we use `sync_external_customer_vm_group` in the sync-loop for cleanup, shouldn't we reset the group members also if `expected_members` is empty?

joker-at-work replied:
Oh. Good point.

seb-kro commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -1273,6 +1276,229 @@ def update_admin_vm_group_membership(self, instance, remove=False):
+            # the internal_workload groups.
+            expected_vm_group_names.add(f"{drs_prefix}internal_workload")
+
+        # Creat missing VmGroup objects so we can create rules for them
```
Just in case you care: typo Creat(e)

seb-kro commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -1273,6 +1276,229 @@ def update_admin_vm_group_membership(self, instance, remove=False):
+        We delete any DRS rules matching the naming scheme but referencing a
+        non-existing HostGroup to keep the cluster clean.
+
+        For each HostGroup there's are 2 rules. Configured with affinity to the
```
Just in case you care: typo: there are
